src/asm/ram32.s

```diff
@@ -0,0 +1,60 @@
+.section .ram32, "ax"
+.code32
+
+ram32_start:
+    # Indicate (via serial) that we are executing out of RAM
+    movw $0x3f8, %dx
+    movb $'R', %al
+    outb %al, %dx
+
+setup_page_tables:
+    # First PML2 entry identity maps [0, 2 MiB)
+    movl $0b10000011, (pml2t) # huge (bit 7), writable (bit 1), present (bit 0)
+    # First PML3 entry points to PML2 table
+    movl $pml2t, %eax
+    orb  $0b00000011, %al # writable (bit 1), present (bit 0)
+    movl %eax, (pml3t)
+    # First PML4 entry points to PML3 table
+    movl $pml3t, %eax
+    orb  $0b00000011, %al # writable (bit 1), present (bit 0)
+    movl %eax, (pml4t)
+
+enable_paging:
+    # Load page table root into CR3
+    movl $pml4t, %eax
+    movl %eax, %cr3
+
+    # Set CR4.PAE (Physical Address Extension)
+    movl %cr4, %eax
+    orb  $0b00100000, %al # Set bit 5
+    movl %eax, %cr4
+    # Set EFER.LME (Long Mode Enable)
+    movl $0xC0000080, %ecx
+    rdmsr
+    orb  $0b00000001, %ah # Set bit 8
+    wrmsr
+    # Set CRO.PG (Paging)
+    movl %cr0, %eax
+    orl  $(1 << 31), %eax
+    movl %eax, %cr0
+
+    # Indicate (via serial) that we have enabled paging
+    movw $0x3f8, %dx
+    movb $'P', %al
+    outb %al, %dx
+
+jump_to_64bit:
+    # We are now in 32-bit compatibility mode. To enter 64-bit mode, we need to
+    # load a 64-bit code segment into our GDT.
+    lgdtl gdt64_ptr
+    # Set CS to a 64-bit segment and jump to 64-bit code.
+    ljmpl $(code64_desc - gdt64_start), $ram64_start
+
+gdt64_ptr:
+    .short gdt64_end - gdt64_start - 1 # GDT length is actually (length - 1)
+    .long gdt64_start
+gdt64_start:
+    .quad 0                                     # First descriptor is null
+code64_desc:
+    .quad (1<<43) | (1<<44) | (1<<47) | (1<<53) # Only these bits do anything
```

Comment:
Should I explain what these bits are?

Reply:
Yeah, good idea!

Reply:
Added more comments, I also set `CS.A=1 CS.R=1 CS.C=1`  (bits 40-42). The SDM isn't clear if these bits are ignored or not in 64-bit mode. [EDK2, for example, has `CS.A=1 CS.R=1 CS.C=0`](https://github.com/tianocore/edk2/blob/3806e1fd139775610d8f2e7541a916c3a91ad989/UefiCpuPkg/ResetVector/Vtf0/Ia16/Real16ToFlat32.asm#L82).